core/include/scipp/core/element/arithmetic.h

```diff
@@ -122,24 +126,52 @@ constexpr auto times = overloaded{
     transform_flags::expect_no_in_variance_if_out_cannot_have_variance,
     [](const auto a, const auto b) { return a * b; }};
 
+// truediv defined as in Python.
 constexpr auto divide = overloaded{
-    divide_types_t{},
+    true_divide_types_t{},
     transform_flags::expect_no_in_variance_if_out_cannot_have_variance,
     [](const auto a, const auto b) {
-      // Integer division is truediv, as in Python 3 and numpy
       if constexpr (std::is_integral_v<decltype(a)> &&
                     std::is_integral_v<decltype(b)>)
-        return static_cast<double>(a) / b;
+        return static_cast<double>(a) / static_cast<double>(b);
       else
         return a / b;
     }};
 
-// mod defined as in Python
-constexpr auto mod = overloaded{
-    arg_list<int64_t, int32_t, std::tuple<int32_t, int64_t>,
-             std::tuple<int64_t, int32_t>>,
-    [](const units::Unit &a, const units::Unit &b) { return a % b; },
-    [](const auto a, const auto b) { return b == 0 ? 0 : ((a % b) + b) % b; }};
+// floordiv defined as in Python. Complementary to mod.
+constexpr auto floor_divide = overloaded{
+    floor_divide_types_t{}, transform_flags::expect_no_variance_arg<0>,
+    transform_flags::expect_no_variance_arg<1>,
+    [](const auto a,
+       const auto b) -> std::common_type_t<decltype(a), decltype(b)> {
+      using std::floor;
+      if constexpr (std::is_integral_v<decltype(a)> &&
```

Comment:
Bit late on the review of this - this could be further cleaned up by defining `numeric::floor` see https://github.com/scipp/scipp/blob/main/common/include/scipp/common/numeric.h for other examples.